pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,450 @@
+		return nil, fmt.Errorf("failed to get alibaba client: %v", err)
+	}
+
+	createInstanceRequest := ecs.CreateCreateInstanceRequest()
+	createInstanceRequest.ImageId, _ = p.getImageIDForOS(machine.Spec, pc.OperatingSystem)
+	createInstanceRequest.InstanceName = machine.Name
+	createInstanceRequest.InstanceType = c.InstanceType
+	createInstanceRequest.VSwitchId = c.VSwitchID
+	createInstanceRequest.InternetMaxBandwidthOut = requests.Integer(c.InternetMaxBandwidthOut)
+	encodedUserData := base64.StdEncoding.EncodeToString([]byte(userdata))
+	createInstanceRequest.UserData = encodedUserData
+	createInstanceRequest.SystemDiskCategory = "cloud_efficiency"
+	createInstanceRequest.ZoneId = c.ZoneID
+
+	_, err = client.CreateInstance(createInstanceRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create instance at Alibaba cloud: %v", err)
+	}
+
+	if err = data.Update(machine, func(updatedMachine *v1alpha1.Machine) {
+		if !kuberneteshelper.HasFinalizer(updatedMachine, finalizerInstance) {
+			updatedMachine.Finalizers = append(updatedMachine.Finalizers, finalizerInstance)
+		}
+	}); err != nil {
+		return nil, fmt.Errorf("failed updating machine %v finzaliers: %v", machine.Name, err)
+	}
+
+	foundInstance, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get alibaba instance %v due to %v", machine.Name, err)
+	}
+
+	tag := ecs.AddTagsTag{
+		Value: string(machine.UID),
+		Key:   machineUIDTag,
+	}
+	request := ecs.CreateAddTagsRequest()
+	request.ResourceType = "instance"
+	request.ResourceId = foundInstance.InstanceId
+	tags := []ecs.AddTagsTag{tag}
+	request.Tag = &tags
+
+	if _, err := client.AddTags(request); err != nil {
+		return nil, fmt.Errorf("failed to create new UID tag: %v", err)
+	}
+
+	return &alibabaInstance{instance: foundInstance}, nil
+}
+
+func (p *provider) Cleanup(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData) (bool, error) {
+	if _, err := p.Get(machine, data); err != nil {
+		if err == cloudprovidererrors.ErrInstanceNotFound {
+			return true, nil
+		}
+		return false, err
+	}
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return false, fmt.Errorf("failed to get alibaba client: %v", err)
+	}
+
+	foundInstance, err := getInstance(client, machine.Name)
+	if err != nil {
+		return false, fmt.Errorf("failed to get alibaba instance %v due to %v", machine.Name, err)
+	}
+
+	deleteInstancesRequest := ecs.CreateDeleteInstancesRequest()
+	deleteInstancesRequest.InstanceId = &[]string{foundInstance.InstanceId}
+
+	deleteInstancesRequest.Force = requests.Boolean("True")
+	if _, err = client.DeleteInstances(deleteInstancesRequest); err != nil {
```

Comment:
Please update this code to ignore 404, i.E. if someone deleted the instance out of bounds, else we are stuck here because we won't remove the finalizer but the deletion doesn't succeed here

Reply:
That's a good catch, however, I will add the check before the deletion. Here: https://github.com/kubermatic/machine-controller/pull/643/files/cc2554f14d210965e8c11b7945929038359302e7#diff-e96854fadd3b15014f29f87ac940a01bR272

Reply:
I see. But then you need to move the finalizer cleanup after the `if err == cloudprovidererrors.ErrInstanceNotFound`